### DISCUSSION TITLE
Client: validate queries against schema?

### DISCUSSION THREAD

**Discussion by pjwalstrom** [graphql-java/graphql-java#2228] [Q&A] [answered]
Hello,
I would like to take all my queries and validate those against a given schema in a unit-test. Is there an easy way to do this?

**Comment by bbakerman** [comment, marked as answer]
Yes.

If your have your queries (in text  form) and your have your schema (in SDL form or runtime form) the you can just run the query against a GraphQL java instance with a MockedRuntimeWiring (one that does nothing) and if it passes back no errors then it must be valid but if it passes back validation errors then it must be invalid as a query.

This can be done in a unit test environment (because running up a  Graphql + GraphqlSchema is unit test cheap - you dont need http etc.. to run graphql queries)

**Reply by pjwalstrom** [reply]
great, thank you! Will try your suggestions!

**Reply by pjwalstrom** [reply]
Ended up with a test like this, @bbakerman. Do not know if it is optimal, but validates as expected. Thanks for the hints!

```java
@Test
public void queriesAreValid() throws IOException {
    Map<String, Object> ident = Collections.singletonMap("ident", "foo");
    TypeDefinitionRegistry typeRegistry = new SchemaParser().parse(Paths.get("src/test/resources/schema.graphqls").toFile());
    GraphQLSchema schema = new SchemaGenerator().makeExecutableSchema(typeRegistry, RuntimeWiring.MOCKED_WIRING);
    GraphQL graphQL = GraphQL.newGraphQL(schema).build();
    Stream<Path> files = Files.list(Paths.get("src/main/resources/queries/"));

    files.forEach(file -> {
        try {
            String query = Files.lines(file).collect(Collectors.joining());
            ExecutionInput executionInput = ExecutionInput.newExecutionInput().query(query).variables(ident).build();
            List<GraphQLError> errors = graphQL.execute(executionInput).getErrors();
            assertTrue("Invalid query: " + file, errors.isEmpty());
        } catch (IOException e) {
            // ignore
        }
    });
}
```

**Reply by bbakerman** [reply]
Yup that is what I meant.  The use of a `RuntimeWiring.MOCKED_WIRING` allows you to create executable queries that dont give any data but can give validation errors

**Reply by pjwalstrom** [reply]
@bbakerman I have a follow-up question, if you don't mind :-)  I have a query that returns a non-null value according to the scema:
```
type Query {
    hentPersonBolk(identer: [ID!]!): [HentPersonBolkResult!]!
}
```
with the following query
```
query myQuery($identer: [ID!]!)  {
    hentPersonBolk(identer: $identer) {
        code
    }
}
```

I get the following error using the mocked wiring:

> The field at path '/hentPersonBolk' was declared as a non null type, but the code involved in retrieving data has wrongly returned a null value.  The graphql specification requires that the parent field be set to null, or if that is non nullable that it bubble up null to its parent and so on. The non-nullable type is '[HentPersonBolkResult!]' within parent type 'Query'

Is this due to the mocked wiring? If that's the case, how could I tell the mocked wiring to return something different than `null`?

**Reply by bbakerman** [reply]
Hmmm good point.  You have moved on beyond the "validation stage" and have started executing the query and MockedWiring will have data fetchers that do nothing I think.

If you want to stop execution (but do validation) then you can throw `AbortExecutionException` in a key place

For example inside `graphql.execution.instrumentation.Instrumentation#beginExecuteOperation` say

That is you would put in a Instrumentation that decides to short cut execution via that exception.

`graphql.analysis.MaxQueryComplexityInstrumentation` is an example of this (albeit it does it on validation and you don't want that)

**Reply by pjwalstrom** [reply]
Thank you, @bbakerman ! Ended up with a bit more code, but a test that validates without executing

```java
    @Test
    public void queriesAreValid() throws IOException {
        Map<String, Object> ident = Collections.singletonMap("ident", Collections.singletonList("42"));

        InputStream inputStream = getClass().getClassLoader().getResourceAsStream("pdl-schema.graphqls");
        TypeDefinitionRegistry typeRegistry = new SchemaParser().parse(inputStream);
        GraphQLSchema schema = new SchemaGenerator().makeExecutableSchema(typeRegistry, RuntimeWiring.MOCKED_WIRING);
        GraphQL graphQL = GraphQL.newGraphQL(schema).instrumentation(new AbortExecutionInstrumentation()).build();

        Files.list(Paths.get("src/main/resources/pdl/queries/")).forEach(file -> {
            try {
                String query = Files.lines(file).collect(Collectors.joining());
                ExecutionInput executionInput = ExecutionInput.newExecutionInput().query(query).variables(ident).build();
                List<GraphQLError> errors = graphQL.execute(executionInput).getErrors();
                if (!queryIsValid(errors)) {
                    fail("Invalid query: " + file + ". " + errors.get(0).getMessage());
                }
            } catch (IOException e) {
                fail(e.getMessage());
            }
        });
    }

    private boolean queryIsValid(List<GraphQLError> notOnlyErrors) {
        // if AbortExecutionException is the only one in the list, everything is ok, as query has gone through validation, but stopped at execution
        return notOnlyErrors.size() == 1 && notOnlyErrors.get(0) instanceof AbortExecutionException;
    }

    private static class AbortExecutionInstrumentation extends SimpleInstrumentation {
        @Override
        public InstrumentationContext<ExecutionResult> beginExecuteOperation(InstrumentationExecuteOperationParameters parameters) {
            throw new AbortExecutionException("We stop before execution");
        }
    }
```

**Reply by bbakerman** [reply]
I have to apologize because I have given you a bum steer.

There is a class that allows parsing and and validating without execution.   It came to me in the shower.

`graphql.ParseAndValidate`

It does all that you want without the execution malarky.

I am sorry to have not told you earlier

**Reply by pjwalstrom** [reply]
Thank you, the ParseAndValidate did exactly what I needed without making an AbortExecutionInstrumentation. I'm happy you had such a great shower thought :-)

**Comment by bbakerman** [comment]
AbortExecutionInstrumentation might be a great thing to add to
graphql-java - as a way (along with MOCK WIRRING) to mock out just for
validation

I guess we dont want the AbortExecutionException induced error
however.  Hmmm need to consider this more
